### PR TITLE
added the ability to edit role names

### LINKED ISSUE #935: Ability to change the name of User Roles
This is an enhancement request.

Allow for the changing of user role names in the Admin module.    Roles are going to correspond to job titles in a hospital ( that is how users will think about them)......and those titles may be unique to a hospital (or country, or region).   Flexibility to change the role name would be very useful.

For the target contexts, I think we have enough roles in the system (15-16?) at this point.  So, while adding and deleting roles *may* be useful, for v1.0 I do not think they are needed.

@jkleinsc

### PR DESCRIPTION
**Pull request by bradgearon**
Fixes #935 

**Changes proposed in this pull request:**
- added input box for role name to admin/roles
- using role id for role select value in view
- using role element in selectRole action / storing selected role id
- added namedRoles map to user roles (filtered roles in user roles controller now uses this)
- updating loaded userRoles on save with new name (using unmodified role id)
- loading user roles and using role name on users index

image

cc @HospitalRun/core-maintainers

### CODE REVIEW

**Comment by bradgearon** [comment]
I just noticed that this needs to be updated on the users page as well, so hold off on merging this please.

**Comment by bradgearon** [comment]
this is working now, and is ready for review! 👍

File: app/admin/roles/controller.js [hunk 1]
```diff
@@ -194,10 +195,13 @@ export default AbstractEditController.extend(UserRoles, UserSession, {
           }
         });
       });
+      roleToUpdate.set('name', currentRole);
       roleToUpdate.set('capabilities', capabilitiesToSave);
       roleToUpdate.save().then(() => {
         this.displayAlert(this.get('i18n').t('admin.roles.titles.roleSaved'),
           this.get('i18n').t('admin.roles.messages.roleSaved', { roleName: currentRole }));
+        let thisRole = this.userRoles.findBy('id', roleId);
+        Ember.set(thisRole, 'name', currentRole);
```
Comment (jkleinsc) [inline app/admin/roles/controller.js]: This line is failing in the tests because currentRole is not set.

File: app/users/index/controller.js [hunk 2]
```diff
@@ -1,8 +1,25 @@
...
-export default AbstractPagedController.extend(UserSession, {
+import UserRoles from 'hospitalrun/mixins/user-roles';
+
+export default AbstractPagedController.extend(UserSession, UserRoles, {
   addPermission: 'add_user',
   deletePermission: 'delete_user',
-  sortProperties: ['displayName']
-
+  sortProperties: ['displayName'],
+  namedRolesChanged: Ember.observer('namedRoles', function() {
```
Comment (jkleinsc) [inline app/users/index/controller.js]: [Observers are an anti-pattern](https://dockyard.com/blog/2015/11/16/best-practices-functional-programming-and-the-observer-effect) and should be avoided.  We are trying to remove as many observers as possible from HospitalRun and I would like to see this refactored/changed to not use an observer.

**Comment by bradgearon** [comment]
@jkleinsc thank you for the review, I've made some changes to resolve those issues

**Comment by jkleinsc** [comment]
@bradgearon thanks for making those changes.  It does look like the tests are still failing:
    ---
not ok 99 PhantomJS 2.1 - Acceptance | roles: visiting /admin/roles
    ---
        actual: >
            The  role has been saved.
        expected: >
            The Doctor role has been saved.
        stack: >

**Comment by bradgearon** [comment]
Oopsie. I'll fix that. Thank you!

**Comment by jkleinsc** [comment]
@bradgearon I see that tests are failing after the last merge.  Let me know if you need any help resolving those errors.

**Comment by tangollama** [comment]
@bradgearon I'm bringing this PR up the latest. If the CI fails again, are you still interested in resolving the error?

**Comment by bradgearon** [comment]
Yeah I'll work on it. Thanks!

**Comment by tangollama** [comment]
@bradgearon, consider this a gentle #bump. I'm going through PR's and am happy to include these changes once the CI checks out.

**Comment by tangollama** [comment]
@bradgearon, if you don't have time to work on this, I could ping the rest of the community to complete this in a new PR.

**Comment by DerekTBrown** [comment]
@tangollama Would be willing to take this over.

**Comment by tangollama** [comment]
@DerekTBrown please do.

**Comment by DerekTBrown** [comment]
Close in favor of https://github.com/HospitalRun/hospitalrun-frontend/pull/1320

**Pull request closed without merging**